File: tomate.c
#include <stdbool.h>
#include <stdlib.h>

/*

Bad copy of Lua pattern matching for strings

Pattern: # ^ $ not implemented
  ^? PatternItem+ $?

PatternItem:
  Class
  Class *
  Class +
  Class -
  Class ?

Class:
  Char
  Any
  LineBreak
  Alphabet
  Digits
  Whitespace
  Lowercase 
  Uppercase
  Alphanum
  Set
*/

#define CLASS_CHAR       0b00000000
#define CLASS_ANY        0b00000001
#define CLASS_BREAK      0b00000010
#define CLASS_LETTER     0b00000011
#define CLASS_DIGITS     0b00000100
#define CLASS_WHITESPACE 0b00000101
#define CLASS_LOWERCASE  0b00000110
#define CLASS_UPPERCASE  0b00000111
#define CLASS_ALPHANUM   0b00001000
#define CLASS_SET        0b00001001
#define CLASS_BITMASK    0b00001111

#define ITEM_SINGLE  0b00000000
#define ITEM_MANY    0b00010000
#define ITEM_MANY1   0b00100000
#define ITEM_SOME    0b00110000
#define ITEM_MAYBE   0b01000000
#define ITEM_BITMASK 0b01110000

#define NEG_BITMASK  0b10000000

struct Item {
    char kind;
    short val;
    short end;
};

struct Pattern {
    bool start;
    bool end;
    char *original;
    int size;
    struct Item *seq;
};
typedef struct Pattern Pattern;

struct MatchResult {
    int start;
    int size;
};
typedef struct MatchResult MatchResult;

MatchResult MATCH_FAILURE = {-1, -1};

bool match_set(char* original, char chr, int start, int end, bool comp) {
    char *iter = &original[start];
    char *iter_end = &original[end];

    for (;iter != iter_end; ++iter) {
        bool cond;

        if (*iter == '%') {
            cond = chr == iter[1];
            iter += 1;
        } else if (iter + 2 < iter_end && iter[1] == '-') {
            cond = chr >= iter[0] && chr <= iter[2];
            iter += 2;
        } else {
            cond = chr == iter[0];
        }

        if (cond) {
            return true;
        }
    }
    return false;
}

bool match_class(struct Item item, char chr, char* original) {
    bool match;
    switch (item.kind & CLASS_BITMASK) {
        case CLASS_CHAR:
            match = chr == item.val;
            break;
        case CLASS_ANY:
            match = chr != '\0';
            break;
        case CLASS_BREAK:
            match = chr == '\n' || chr == '\r';
            break;
        case CLASS_LETTER:
            match = (chr >= 'a' && chr <= 'z') 
                 || (chr >= 'A' && chr <= 'Z');
            break;
        case CLASS_DIGITS:
            match = chr >= '0' && chr <= '9';
            break;
        case CLASS_WHITESPACE:
            match = chr == ' ' || chr == '\t' || chr == '\n' || chr == '\r';
            break;
        case CLASS_LOWERCASE:
            match = (chr >= 'a' && chr <= 'z');
            break;
        case CLASS_UPPERCASE:
            match = (chr >= 'A' && chr <= 'Z');
            break;
        case CLASS_ALPHANUM:
            match = (chr >= 'a' && chr <= 'z') 
                 || (chr >= 'A' && chr <= 'Z')
                 || (chr >= '0' && chr <= '9')
                 || chr == '_';
            break;
        case CLASS_SET:
            match = match_set(original, chr, item.val, item.end,
                (item.kind & NEG_BITMASK) != 0);
            break;
    }
    return ((item.kind & NEG_BITMASK) != 0) != match;
}

int match_item(
    struct Item *seq, int seq_size, int seq_index, 
    char *str, int str_size, int str_index,
    char *original
) {
    int size;
    int last_size;
    char *iter = str;
    struct Item item = seq[seq_index];
    switch (item.kind & ITEM_BITMASK) {
        case ITEM_SINGLE:
            if (match_class(item, *iter, original)) {
                return 1;
            }
            return -1;
        case ITEM_MANY:
        case ITEM_MANY1:
            size = 0;
            last_size = size;
            while (match_class(item, *iter, original)) {
                ++size;
                ++iter;

                if (seq_index + 1 >= seq_size || 
                    match_class(seq[seq_index+1], *iter, original))
                {
                    last_size = size;
                }
            }

            if (item.kind == ITEM_MANY1 && size == 0) {
                return -1;
            }
            return last_size;
        case ITEM_SOME:
            size = 0;
            while (match_class(item, *iter, original)) {
                ++size;
                ++iter;

                if (seq_index + 1 < seq_size && 
                    match_class(seq[seq_index+1], *iter, original))
                {
                    break;
                }
            }
            return size;
        case ITEM_MAYBE:
            if (match_class(item, *iter, original)) {
                return 1;
            }
            return 0;
    }
    return -1;
}

MatchResult match(Pattern pat, char *str, int str_size, int match_start) {
    if (match_start >= str_size) return MATCH_FAILURE;

    char *ptr = &str[match_start];
    if (*ptr == '\0') return MATCH_FAILURE;

    int pat_index = 0;

    int size = 0;

    while (pat_index < pat.size) {
        int res_size = match_item(
            pat.seq, pat.size, pat_index,
            ptr, str_size-match_start, match_start,
            pat.original
        );

        if (res_size == -1) {
            return MATCH_FAILURE;
        }

        size += res_size;
        ptr += res_size;

        pat_index += 1;
    }

    return (MatchResult){ match_start, size };
}

struct Item pchar(char *str, int *len, int index) {
    #define X (struct Item)

    if (*str == '.') {
        if (len) *len = 1;
        return X {CLASS_ANY};
    }

    if (*str == '[') {
        bool negate = false;
        int start = index + 1;
        int end = start;
        if (str[1] == '^') {
            negate = true;
            start += 1;
            end += 1;
        }
        while (str[end] != '\0') {
            if (str[end] == '%') {
                end += 1;
            }
            if (str[end] == ']') {
                break;
            }
            end += 1;
        }
        if (len) *len = end - index + 1;
        return X { CLASS_SET | (negate ? NEG_BITMASK : 0), start, end};
    }

    if (*str != '%') {
        if (len) *len = 1;
        return X {.val = *str};
    }

    if (len) *len = 2;
        
    switch (str[1]) {
        case 'n': return X { CLASS_BREAK };
        case 'N': return X { CLASS_BREAK | NEG_BITMASK };
        case 'a': return X { CLASS_LETTER };
        case 'A': return X { CLASS_LETTER | NEG_BITMASK };
        case 'd': return X { CLASS_DIGITS };
        case 'D': return X { CLASS_DIGITS | NEG_BITMASK };
        case 's': return X { CLASS_WHITESPACE };
        case 'S': return X { CLASS_WHITESPACE | NEG_BITMASK };
        case 'l': return X { CLASS_LOWERCASE };
        case 'u': return X { CLASS_UPPERCASE };
        case 'w': return X { CLASS_ALPHANUM };
        case 'W': return X { CLASS_ALPHANUM | NEG_BITMASK };
        default:
            return X {.val = str[1]};
    }

    #undef X
}

struct Item pitem(char *str, int *len, int index) {
    int len_;
    struct Item c = pchar(str, &len_, index);

    if (len) *len = len_ + 1;

    switch (str[len_]) {
        case '*': c.kind |= ITEM_MANY; break;
        case '+': c.kind |= ITEM_MANY1; break;
        case '-': c.kind |= ITEM_SOME; break;
        case '?': c.kind |= ITEM_MAYBE; break;
        default:
            if (len) *len = *len - 1;
            c.kind |= ITEM_SINGLE;
            break;
    }

    return c;
}

Pattern compile(char *str) {
    struct Item *buffer = NULL;
    int buffer_size = 0;
    int buffer_capacity = 8;

    char *original = str;

    int len_;
    int index = 0;
    buffer = realloc(buffer, buffer_capacity * sizeof(struct Item));

    while (*str != '\0') {
        struct Item item = pitem(str, &len_, index);
        str += len_;
        index += len_;

        buffer[buffer_size] = item;

        if (buffer_size + 1 == buffer_capacity) {
            buffer_capacity += 8;
            buffer = realloc(buffer, buffer_capacity * sizeof(struct Item));
        }

        ++buffer_size;
    }

    Pattern pat = {
        false, false, original, buffer_size, buffer
    };

    return pat;
}

File: go/tomate/tomate.go
package tomate

type charClassKind int
type patternItemKind int

const (
	classChar charClassKind = iota
	classAny
	classBreak
	classLetter
	classDigits
	classWhitespace
	classLowercase
	classUppercase
	classAlphanum
	classSet
)

const (
	itemSingle patternItemKind = iota
	itemMany
	itemMany1
	itemSome
	itemMaybe
)

type patternItem struct {
	class charClassKind
	item patternItemKind
	negate bool
	val rune
	start int
	end int
}

type Pattern struct {
	original string
	seq []patternItem
}

type MatchResult struct {
	start int
	end int
}

func matchSet(char rune, original string, start int, end int) bool {
    var set = []rune(original[start:end])

    for len(set) > 0 {
        var cond bool

        if set[0] == '%' {
            cond = char == set[1]
            set = set[1:]
        } else if len(set) >= 3 && set[1] == '-' {
            cond = char >= set[0] && char <= set[2]
            set = set[2:]
        } else {
            cond = char == set[0]
        }

        if cond  {
            return true
        }

        set = set[1:]
    }
    return false
}

func matchClass(item patternItem, char rune, original string) bool {
	var match bool
	switch item.class {
	case classChar:
		match = char == item.val
	case classAny:
		match = char != '\x00'
	case classBreak:
		match = char == '\n' || char == '\r'
	case classLetter:
		match = (char >= 'a' && char <= 'z') || (char >= 'A' && char <= 'Z')
	case classDigits:
		match = char >= '0' && char <= '9'
	case classWhitespace:
		match = char == ' ' || char == '\t' || char == '\n' || char == '\r'
	case classLowercase:
		match = (char >= 'a' && char <= 'z')
	case classUppercase:
		match = (char >= 'A' && char <= 'Z')
	case classAlphanum:
		match = 
			(char >= 'a' && char <= 'z') ||
			(char >= 'A' && char <= 'Z') || 
			(char >= '0' && char <= '9') || 
			char == '_'

	case classSet:
	    match = matchSet(char, original, item.start, item.end)
	}
	return item.negate != match;
}

func matchItem(seq []patternItem, seqIndex int, str []rune, original string) int {
	var iter = str

	var item = seq[seqIndex]

	switch item.item {
	case itemSingle:
		if len(iter) > 0 && matchClass(item, iter[0], original) {
			return 1
		}
		return -1
	case itemMany, itemMany1:
		size := 0

		for len(iter) > 0 && matchClass(item, iter[0], original) {
			size += 1
			iter = iter[1:]
		}

		if item.item == itemMany1 && size == 0 {
			return -1;
		}

		return size
	case itemSome:
		size := 0
		for len(iter) > 0 && matchClass(item, iter[0], original) {
			size += 1
			iter = iter[1:]

			if seqIndex + 1 < len(seq) && len(iter) > 0 && matchClass(seq[seqIndex+1], iter[0], original) {
				break
			}
		}
		return size
	case itemMaybe:
		if len(iter) > 0 && matchClass(item, iter[0], original) {
			return 1
		}
		return 0
	}
	return -1
}

func Match(pat Pattern, str string, matchStart int) (bool, MatchResult) {
	var matchFailure = MatchResult { -1, -1 }

    var strSlice []rune = []rune(str)
    var strIter = matchStart

    if matchStart >= len(str) {
    	return false, matchFailure
    }

    var patIter = 0
    var size = 0

    for ;patIter < len(pat.seq); patIter += 1 {
        var resultSize = matchItem(pat.seq, patIter, strSlice[strIter:], pat.original)

        if resultSize == -1 {
            return false, MatchResult { matchStart, size }
        }

        size += resultSize
        strIter += resultSize
    }

    return true, MatchResult { matchStart, size }
}

func parseChar(str []rune, index int) (patternItem, int) {

	if str[0] == '.' {
		return patternItem { class: classAny }, 1
	}

    if (str[0] == '[') {
    	negate := false
        
        start :=  1
        end := start
        
        if str[1] == '^' {
            negate = true
            start += 1
            end += 1
        }
        for end < len(str) {
            if str[end] == '%' {
                end += 1
            }
            if str[end] == ']' {
                break;
            }
            end += 1
        }

        return patternItem {
        	class: classSet,
        	negate: negate,
        	start: index + start,
        	end: index + end,
        }, end - index + 1
    }

    if str[0] != '%' {
        return patternItem { class: classChar, val: str[0] }, 1
    }
        
    switch str[1] {
    case 'n': return patternItem { class: classBreak }, 2
    case 'N': return patternItem { class: classBreak, negate: true }, 2
    case 'a': return patternItem { class: classLetter }, 2
    case 'A': return patternItem { class: classLetter, negate: true }, 2
    case 'd': return patternItem { class: classDigits }, 2
    case 'D': return patternItem { class: classDigits, negate: true }, 2
    case 's': return patternItem { class: classWhitespace }, 2
    case 'S': return patternItem { class: classWhitespace, negate: true }, 2
    case 'l': return patternItem { class: classLowercase }, 2
    case 'u': return patternItem { class: classUppercase }, 2
    case 'w': return patternItem { class: classAlphanum }, 2
    case 'W': return patternItem { class: classAlphanum, negate: true }, 2
    default:  return patternItem { class: classChar, val: str[1] }, 2
    }
}

func parseItem(str []rune, index int) (patternItem, int) {
    item, parseLen := parseChar(str, index)

    if parseLen >= len(str) {
    	item.item = itemSingle
    	return item, parseLen
    }

    switch str[parseLen] {
        case '*': item.item = itemMany
        case '+': item.item = itemMany1
        case '-': item.item = itemSome
        case '?': item.item = itemMaybe
        default:
            item.item = itemSingle
            return item, parseLen
    }

    return item, parseLen + 1
}

func Compile(str string) Pattern {
    var buffer []patternItem = nil
    var strIter = []rune(str)

    var index = 0

    for len(strIter) > 0 {
        item, len := parseItem(strIter, index)

        strIter = strIter[len:]
        index += len

        buffer = append(buffer, item)
    }

    return Pattern {
    	original: str,
    	seq: buffer,
    }
}
